crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,280 @@
+use crate::nvtxw_bindings::NVTXW3_INTERFACE_ID_CORE_V1;
+use crate::nvtxw_bindings::NVTXW3_RESULT_FAILED;
+use crate::nvtxw_bindings::NVTXW3_RESULT_INVALID_ARGUMENT;
+use crate::nvtxw_bindings::NVTXW3_RESULT_SUCCESS;
+use crate::nvtxw_bindings::NVTXW3_STREAM_ORDERING_SKID_NONE;
+
+pub const NVTX_PAYLOAD_ENTRY_FLAG_POINTER: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_POINTER as u64;
+pub const NVTX_PAYLOAD_ENTRY_FLAG_EVENT_MESSAGE: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_EVENT_MESSAGE as u64;
+pub const NVTX_PAYLOAD_ENTRY_FLAG_EVENT_TIMESTAMP: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_EVENT_TIMESTAMP as u64;
+pub const NVTX_PAYLOAD_ENTRY_FLAG_RANGE_BEGIN: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_RANGE_BEGIN as u64;
+pub const NVTX_PAYLOAD_ENTRY_FLAG_RANGE_END: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_RANGE_END as u64;
+pub const NVTX_PAYLOAD_ENTRY_TYPE_UINT64: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_TYPE_UINT64 as u64;
+pub const NVTX_PAYLOAD_ENTRY_TYPE_NVTX_COLOR: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_TYPE_NVTX_COLOR as u64;
+pub const NVTX_PAYLOAD_ENTRY_TYPE_CSTRING: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_TYPE_CSTRING as u64;
+pub const NVTX_PAYLOAD_SCHEMA_TYPE_STATIC: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_TYPE_STATIC as u64;
+pub const NVTX_PAYLOAD_SCHEMA_FLAG_NONE: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_FLAG_NONE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_NAME: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_NAME as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_TYPE: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_TYPE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_FLAGS: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_FLAGS as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_ENTRIES: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID as u64;
+
+pub const NVTXW3_STREAM_ORDER_INTERLEAVING_NONE: i16 = 0;
+pub const NVTXW3_STREAM_ORDERING_TYPE_UNKNOWN: i16 = 0;
+
+pub fn initialize_simple() -> Result<InterfaceHandle, ResultCode> {
+    let so_name = "libNvtxwBackend.so";
+    let c_mode_string = CString::new(so_name).expect("modeString CString::new failed");
+    let cptr_mode_string = c_mode_string.as_ptr();
+
+    let result;
+    let mut get_interface_func: nvtxw_bindings::nvtxwGetInterface_t = Default::default();
+    unsafe {
+        result = nvtxw_bindings::nvtxwInitialize(
+            NVTXW3_INIT_MODE_LIBRARY_FILENAME,
+            cptr_mode_string,
+            &mut get_interface_func,
+            null_mut(),
+        );
+    }
+
+    if result != NVTXW3_RESULT_SUCCESS {
+        return Err(result);
+    }
```

Comment:
Bonus points for marking the `nvtxwInitialize` result as `must_use`: https://doc.rust-lang.org/std/hint/fn.must_use.html (and so on for other APIs).

Reply:
Thanks for the suggestion, the code is a lot cleaner this way.

For now, I'll hold off on `must_use` because it is still experimental. Once it is stable I can add it.